curvefs/src/metaserver/storage/rocksdb_storage.cpp

```diff
@@ -473,27 +366,140 @@ Status RocksDBStorage::Rollback()  {
+
+const char* const kRocksdbCheckpointPath = "rocksdb_checkpoint";
 
-    if (!GetProcMemory(&statistics->memoryUsageBytes)) {
-        LOG(ERROR) << "Get process memory failed.";
+bool DoCheckpoint(rocksdb::DB* db, const std::string& dest) {
+    rocksdb::Checkpoint* ckptr = nullptr;
+    auto status  = rocksdb::Checkpoint::Create(db, &ckptr);
+    if (!status.ok()) {
+        LOG(ERROR) << "Failed to create checkpoint, " << status.ToString();
         return false;
     }
 
-    uint64_t total, available;
-    if (!GetFileSystemSpaces(options_.dataDir, &total, &available)) {
-        LOG(ERROR) << "Get filesystem space failed.";
+    std::unique_ptr<rocksdb::Checkpoint> ckptrGuard(ckptr);
+    status = ckptr->CreateCheckpoint(dest);
+    if (!status.ok()) {
+        LOG(ERROR) << "Failed to create checkpoint at `" << dest << "`, "
+                   << status.ToString();
         return false;
     }
-    statistics->diskUsageBytes = total - available;
 
     return true;
 }
 
-StorageOptions RocksDBStorage::GetStorageOptions() const {
-    return options_;
+}  // namespace
+
+bool RocksDBStorage::Checkpoint(const std::string& dir,
+                                std::vector<std::string>* files) {
+    rocksdb::FlushOptions options;
+    options.wait = true;
+    options.allow_write_stall = true;
+    auto status = db_->Flush(options, handles_);
+    if (!status.ok()) {
+        LOG(ERROR) << "Failed to flush DB, " << status.ToString();
+        return false;
+    }
+
+    const std::string dest = dir + "/" + kRocksdbCheckpointPath;
+    if (!DoCheckpoint(db_, dest)) {
+        return false;
+    }
+
+    std::vector<std::string> filenames;
+    int ret = options_.localFileSystem->List(dest, &filenames);
+
+    if (ret != 0) {
+        LOG(ERROR) << "Failed to list checkpoint files at `" << dest << "`, "
+                   << berror();
+        return false;
+    }
+
+    files->reserve(filenames.size());
+    for (const auto& f : filenames) {
+        files->push_back(std::string(kRocksdbCheckpointPath) + "/" + f);
+    }
+
+    return true;
+}
+
+bool DuplicateRocksdbCheckpoint(const std::string& from,
+                                const std::string& to) {
+    LOG(INFO) << "Duplicating rocksdb storage from `" << from << "` to `" << to
+              << "`";
+
+    rocksdb::DB* db = nullptr;
+    rocksdb::DBOptions dbOptions;
+    std::vector<rocksdb::ColumnFamilyDescriptor> columnFamilies;
+
+    InitRocksdbOptions(&dbOptions, &columnFamilies);
+
+    std::vector<rocksdb::ColumnFamilyHandle*> cfHandles;
+
+    auto status = rocksdb::DB::OpenForReadOnly(
+        dbOptions, from, columnFamilies, &cfHandles, &db,
+        /* error_if_wal_file_exists */ true);
+
+    if (!status.ok()) {
+        LOG(ERROR) << "Failed to open checkpoint, error: " << status.ToString();
+        return false;
+    }
+
+    std::unique_ptr<rocksdb::DB> dbGuard(db);
+    for (auto* handle : cfHandles) {
+        status = db->DestroyColumnFamilyHandle(handle);
```

Comment:
Why `DestroyColumnFamilyHandle` before reading

Reply:
it's equal to `delete handle`, and we don't read from the database, just do a checkpoint.